### PR TITLE
Support DataFrame sum

### LINKED ISSUE #575: Implement Series sum
**Is your feature request related to a problem? Please describe.**
As a first step to implement reduction operators for DataFrame, we can implement sum for Series as a initial support.

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Support sum for DataFrame as a initial support for reduction operators.

- [x] Series.sum

- [x] DataFrame.sum

## Related issue number

Resolves #575

### CODE REVIEW

File: mars/dataframe/reduction/core.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+        while len(chunks) > 1:
+            new_chunks = []
+            for i in range(0, len(chunks), combine_size):
+                chks = chunks[i: i + combine_size]
+                concat_op = DataFrameConcat(object_type=ObjectType.series)
+                length = sum([c.shape[0] for c in chks if len(c.shape) > 0])
+                chk = concat_op.new_chunk(chks, shape=(length,), index=(i,), dtype=chks[0].dtype,
+                                          index_value=parse_index(pd.RangeIndex(length)))
+                reduced_shape = cls.get_reduced_df(chk, axis=op.axis, level=op.level).shape
```
Comment (qinxuye) [inline mars/dataframe/reduction/core.py]: Same as comments above, no need to calculate meta for every chunk.

File: mars/dataframe/reduction/core.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+    @classmethod
+    def single_tiles(cls, op):
+        df = op.outputs[0]
+        chk = op.inputs[0].chunks[0]
+        reduced_df = cls.get_reduced_df(chk, axis=op.axis, level=op.level)
+        new_chunk_op = op.copy().reset_key()
+        index_value = getattr(reduced_df, 'index', None)
+        if index_value is None:
+            index_value = pd.RangeIndex(0)
```
[the remark this answers is not in the document]
Comment (hekaisheng) [reply mars/dataframe/reduction/core.py]: If `reduced_df` is a scalar.

File: mars/dataframe/reduction/core.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        in_df = op.inputs[0]
+        combine_size = op.combine_size or options.combine_size
+
+        if len(in_df.chunks) == 1:
+            return cls._tile_one_chunk(op)
+
+        n_rows, n_cols = in_df.chunk_shape
+        reduction_chunks = cls._build_reduction_chunks(op)
+        out_chunks = []
+        if op.axis is None or op.axis == 0 or op.axis == 'index':
```
Comment (qinxuye) [inline mars/dataframe/reduction/core.py]: `op.axis` should be written as 0 if it's `index`, 1 if it's `columns` when calling `__call__` to reduce the code.

File: mars/dataframe/reduction/core.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+                new_op._object_type = ObjectType.dataframe
+                new_chunks.append(new_op.new_chunk([chk], shape=reduced_shape, index=(i,), dtypes=dtypes,
+                                                   index_value=index_value))
+            chunks = new_chunks
+        concat_op = DataFrameConcat(axis=op.axis, object_type=ObjectType.dataframe)
+        chk = concat_op.new_chunk(chunks, index=(idx,))
+        empty_df = build_empty_df(chunks[0].dtypes)
+        reduced_df = getattr(empty_df, getattr(cls, '_func_name'))(axis=op.axis, level=op.level,
+                                                                   numeric_only=op.numeric_only)
+        reduced_shape = (np.nan,) if op.axis == 1 or op.axis == 'columns' else reduced_df.shape
```
Comment (qinxuye) [inline mars/dataframe/reduction/core.py]: When axis == 1, the reduced_df.shape is (0, x) which is wrong.

File: mars/dataframe/reduction/core.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        new_op = op.copy().reset_key()
+        return new_op.new_chunk([chk], shape=reduced_shape, index=(idx,), dtype=reduced_df.dtype,
+                                index_value=parse_index(reduced_df.index))
+
+    @classmethod
+    def execute(cls, ctx, op):
+        kwargs = dict(axis=op.axis, level=op.level, skipna=op.skipna, numeric_only=op.numeric_only)
+        in_df = ctx[op.inputs[0].key]
+        res = getattr(in_df, getattr(cls, '_func_name'))(**kwargs)
```
Comment (qinxuye) [inline mars/dataframe/reduction/core.py]: Just pass kwargs here, no need to define it first.

File: mars/dataframe/reduction/tests/test_reduction_execute.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import pandas as pd
+import numpy as np
+
+from mars.tests.core import TestBase
+from mars.dataframe.datasource.series import from_pandas as from_pandas_series
+from mars.dataframe.datasource.dataframe import from_pandas as from_pandas_df
+
+
+class Test(TestBase):
+    def testSeriesSum(self):
```
Comment (qinxuye) [inline mars/dataframe/reduction/tests/test_reduction_execute.py]: `numeric_only` and `min_count` is not tested.

File: mars/dataframe/reduction/tests/test_reduction_execute.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+        data = pd.Series(np.random.rand(20), name='a')
+        data[data > 0.5] = np.nan
+        sum_df = from_pandas_series(data, chunk_size=3).sum()
+        self.assertAlmostEqual(data.sum(), sum_df.execute())
+
+        sum_df = from_pandas_series(data, chunk_size=3).sum(skipna=False)
+        self.assertTrue(np.isnan(sum_df.execute()))
+
+    def testDataFrameSum(self):
```
Comment (qinxuye) [inline mars/dataframe/reduction/tests/test_reduction_execute.py]: `min_count` is not tested.

File: mars/dataframe/reduction/tests/test_reduction_execute.py (hunk 6 above)
Comment (hekaisheng) [reply mars/dataframe/reduction/tests/test_reduction_execute.py]: `numeric_only` only for `DataFrame`.

File: mars/dataframe/reduction/core.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    @property
+    def dtype(self):
+        return self._dtype
+
+    @property
+    def combine_size(self):
+        return self._combine_size
+
+
+class ReductionMixin(DataFrameOperandMixin):
```
Comment (qinxuye) [inline mars/dataframe/reduction/core.py]: Should be renamed to `_DataFrameReductionMixin`.

File: mars/dataframe/reduction/core.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    @classmethod
+    def _get_reduced_meta(cls, chunk, axis):
```
Comment (qinxuye) [inline mars/dataframe/reduction/core.py]: I think this function can be removed. It's signature is different according to different situation, it's hard to maintain, and I think it's unnecessary to calculate meta for each dataframe, series and chunks.

File: .travis.yml [hunk 10]
```diff
@@ -50,7 +50,7 @@ before_install:
   - if [[ "$PYTHON" =~ "2" ]]; then
       pip install --upgrade pip\<19 setuptools\<40.7 wheel coverage;
     else
-      pip install --upgrade pip setuptools wheel coverage;
+      pip install --upgrade pip\<19.2.2 setuptools wheel coverage;
```
Comment (wjsi) [inline .travis.yml]: Remove this workaround as it has already been fixed

File: bin/travis-install-conda.sh [hunk 11]
```diff
@@ -18,6 +18,9 @@ fi
 
 curl -s -o miniconda.sh https://repo.continuum.io/miniconda/$CONDA_FILE
 bash miniconda.sh -b -p $HOME/miniconda && rm miniconda.sh
+if [[ "$PYTHON" == "3.7" ]]; then
+  PYTHON="3.7.3"
+fi
```
Comment (wjsi) [inline bin/travis-install-conda.sh]: ditto

File: mars/dataframe/__init__.py [hunk 12]
```diff
@@ -18,7 +18,11 @@
 # do imports to register operands
 from .arithmetic import add
 from .datasource.from_tensor import from_tensor
+from .reduction import sum_series, sum_dataframe
+del sum_series, sum_dataframe
```
[the remark this answers is not in the document]
Comment (hekaisheng) [reply mars/dataframe/__init__.py]: Do you have any suggestions? Maybe we need a unified manner to register operands.

**Pull request merged**